Add support for eval command to support watch variables

grgustaf opened the pull request:
Caveat: JerryScript only returns a string representation of the eval
  result, so the patch acts like everything is a string. CDT would
  prefer to get typed information and progressively disclose object
  contents, for example, but this would require JrS enhancments.
Caveat: Ignores script parsing messages during eval, but this may need
  to be restored later to allow debugging inside eval'd code.
Chrome sometimes issues eval request while code is running, which
  JerryScript doesn't allow, so they are queued up in that case.
Add unit tests for nearly all of the added code.

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf committed:
Add support for eval command to support watch variables

Caveat: JerryScript only returns a string representation of the eval
  result, so the patch acts like everything is a string. CDT would
  prefer to get typed information and progressively disclose object
  contents, for example, but this would require JrS enhancements.
Caveat: Ignores script parsing messages during eval, but this may need
  to be restored later to allow debugging inside eval'd code.
Chrome sometimes issues eval request while code is running, which
  JerryScript doesn't allow, so they are queued up in that case.
Add unit tests for nearly all of the added code.

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

Signed-off-by: Geoff Gustafson <geoff@linux.intel.com>

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

martijnthe commented on jerry-debugger/src/lib/__tests__/protocol-handler.test.ts:
```diff
@@ -171,6 +182,30 @@ describe('onSourceCodeName', () => {
 });
 
+describe('releaseFunction', () => {
+  it('', () => {
```
Missing a name? `'works'` ?

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -466,6 +542,30 @@ export class JerryDebugProtocolHandler {
 
+  evaluate(expression: string) {
+    if (!this.lastBreakpointHit) {
+      this.evalQueue.push(expression);
```
> Chrome sometimes issues eval request while code is running, which JerryScript doesn't allow, so they are queued up in that case.

What are some examples of the source that Chrome is trying to send? When does this happen?
I assume it always uses the top / global scope then?

To me the queuing feels out of place here because it's a protocol parsing concern.
I think it's better to put it into the `cdt-controller.ts`.

grgustaf replied:
If you set up watch expressions, it tries to evaluate them after each step command. However, it doesn't wait to receive a 'paused' event before doing so - I'm not quite sure why. But JrS doesn't allow that.

Hmm, looking more closely at the trace I guess it calls Debugger.evaluate() on the watch expressions right after the 'resume' event that comes in response to a step command. 

Then, after the next breakpoint is hit (i.e. step is finished), it calls Runtime.evaluateOnCallFrame with call frame of 0. So perhaps I could ignore Debugger.evaluate() for now and only respond to the call frame one. Also, I realize it would probably be better to report failure if they try to evaluate on a call frame other than 0, rather than just evaluate on frame 0 like now.

On relocating to cdt-controller.ts, I remember considering it but not giving it much thought, so you're probably right. I'll try to think about it a bit more. Or maybe given the above I can skip the queuing for now.

grgustaf replied:
After trying to ignore call frames other than 0, this is not actually great because it makes all expressions show up as <not available> even if they were defined at or above the requested call frame. So dropping that thought.

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -410,22 +462,46 @@ export class JerryDebugProtocolHandler {
 
+  onEvalResult(data: Uint8Array) {
+    this.logPacket('Eval Result');
+    this.evalResultData = assembleUint8Arrays(this.evalResultData, data);
```
Why is not a local variable for `this.evalResultData`? I don't see it's being used really, aside of the tests that check that it's set back to `undefined`.

grgustaf replied:
Like most of these other JrS debugger packets, a long result is spread over multiple packets: _RESULTs finalized by _RESULT_END. So the half-baked data persists across calls in that case. If there are static variables, I guess it could use that.

Interesting, the .html client at least doesn't handle this right and always treats them as 'final' packets. Unless it is *I* who am mistaken. :)

martijnthe replied:
Ah, right, sorry I missed that.

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -514,6 +614,12 @@ export class JerryDebugProtocolHandler {
+  logPacket(description: string, ignorable: boolean = false) {
+    // certain packets are ignored while evals are pending
+    const ignored = (ignorable && this.evalsPending) ? 'Ignored: ' : '';
+    console.log(`[${ignored}${description}]`);
```
Hook up the `--verbose` option that's already being parsed in `cli.ts`?

grgustaf replied:
Yeah, it's on the TODO list. :)

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -94,14 +96,19 @@ export class JerryDebugProtocolHandler {
+  private evalsPending: number = 0;
   private lastBreakpointHit?: Breakpoint;
   private lastBreakpointExact: boolean = true;
+  private lastBacktrace?: Array<Breakpoint>;
```
What's this for? I don't see it being used?

grgustaf replied:
You're right, vestigial... I thought I was going to need it to look up call frame info, but there's no way to use it anyway. I can only evaluate on call frame 0. I thought I got rid of all this. :)

martijnthe commented on jerry-debugger/src/lib/protocol-handler.ts:
```diff
@@ -94,14 +96,19 @@ export class JerryDebugProtocolHandler {
   private sourceNameData?: Uint8Array;
   private functionName?: string;
   private functionNameData?: Uint8Array;
+  private evalResult?: string;
```
What's this for? I don't see it being used?

grgustaf replied:
Dropped in latest fixup.

grgustaf committed:
clean up unused variables

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf committed:
add missing test description

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf committed:
Remove support for evaluate() call and eval queuing

For reasons not yet understood, CDT issues a Debugger.evaluate() request
for each watch expresssion after the debugger resumes to execute a
step command. After it hits the next breakpoint, it issues a similar
Runtime.evaluateOnCallFrame() request for each watch expression. JrS can
only do eval while at a breakpoint, so to handle the former we need to
queue them up. Better just to ignore them and handle the latter instead,
without queuing, which seems to work fine.

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

martijnthe commented:
> For reasons not yet understood, CDT issues a Debugger.evaluate() request
for each watch expresssion after the debugger resumes to execute a
step command. After it hits the next breakpoint, it issues a similar
Runtime.evaluateOnCallFrame() request for each watch expression. JrS can
only do eval while at a breakpoint, so to handle the former we need to
queue them up. Better just to ignore them and handle the latter instead,
without queuing, which seems to work fine.

Odd, but glad it works as well with `Runtime.evaluateOnCallFrame()`.

LGTM!

martijnthe commented:
LGTM, feel free to merge at your convenience!
